internal/fwserver/server_moveresourcestate_test.go

```diff
@@ -812,6 +824,148 @@ func TestServerMoveResourceState(t *testing.T) {
+					"test_id": "test_id_value",
+				}),
+				SourceRawState: testNewRawState(t, map[string]interface{}{
+					"id":                 "test-id-value",
+					"required_attribute": true,
+				}),
+				TargetResource: &testprovider.ResourceWithMoveState{
+					MoveStateMethod: func(ctx context.Context) []resource.StateMover {
+						return []resource.StateMover{
+							{
+								StateMover: func(_ context.Context, req resource.MoveStateRequest, resp *resource.MoveStateResponse) {
+									expectedSourceIdentity := testNewRawState(t, map[string]interface{}{
+										"test_id": "test_id_value",
+									})
+
+									if diff := cmp.Diff(req.SourceIdentity, expectedSourceIdentity); diff != "" {
+										resp.Diagnostics.AddError("Unexpected req.SourceIdentity difference", diff)
+									}
+
+									resp.Diagnostics.Append(resp.TargetIdentity.SetAttribute(ctx, path.Root("test_id"), "test_id_value")...)
+
+									// Prevent missing implementation error, the values do not matter except for response assertion
+									resp.Diagnostics.Append(resp.TargetState.SetAttribute(ctx, path.Root("id"), "test-id-value")...)
+									resp.Diagnostics.Append(resp.TargetState.SetAttribute(ctx, path.Root("required_attribute"), "true")...)
+								},
+							},
+						}
+					},
+				},
+				TargetResourceSchema: testSchema,
+				TargetTypeName:       "test_resource",
+				IdentitySchema:       testIdentitySchema,
+			},
+			expectedResponse: &fwserver.MoveResourceStateResponse{
+				TargetPrivate: privatestate.EmptyData(ctx),
+				TargetState: &tfsdk.State{
+					Raw: tftypes.NewValue(schemaType, map[string]tftypes.Value{
+						"id":                 tftypes.NewValue(tftypes.String, "test-id-value"),
+						"optional_attribute": tftypes.NewValue(tftypes.String, nil),
+						"required_attribute": tftypes.NewValue(tftypes.String, "true"),
+					}),
+					Schema: testSchema,
+				},
+				TargetIdentity: &tfsdk.ResourceIdentity{
+					Raw: tftypes.NewValue(schemaIdentityType, map[string]tftypes.Value{
+						"test_id": tftypes.NewValue(tftypes.String, "test_id_value"),
+					}),
+					Schema: testIdentitySchema,
+				},
+			},
+		},
+		"response-invalid-identity": {
+			server: &fwserver.Server{
+				Provider: &testprovider.Provider{},
+			},
+			request: &fwserver.MoveResourceStateRequest{
+				SourceRawState: testNewRawState(t, map[string]interface{}{
+					"id":                 "test-id-value",
+					"required_attribute": true,
+				}),
+				TargetResource: &testprovider.ResourceWithMoveState{
+					MoveStateMethod: func(ctx context.Context) []resource.StateMover {
+						return []resource.StateMover{
+							{
+								StateMover: func(_ context.Context, req resource.MoveStateRequest, resp *resource.MoveStateResponse) {
+									// Try to set TargetIdentity even though req.IdentitySchema is nil
+									resp.Diagnostics.Append(resp.TargetIdentity.SetAttribute(ctx, path.Root("test_id"), "should-not-be-set")...)
```

Comment:
This is the scenario we're trying to prevent, if I do this it'll send back identity data to Terraform despite us not "announcing" an identity schema 🙂:

```suggestion
									// This resource doesn't indicate identity support (via a schema), so this should raise a diagnostic.
									resp.TargetIdentity = &tfsdk.ResourceIdentity{
										Raw: tftypes.NewValue(testIdentitySchema.Type().TerraformType(ctx), map[string]tftypes.Value{
											"test_id": tftypes.NewValue(tftypes.String, "should-not-be-set"),
										}),
										Schema: testIdentitySchema,
									}
```

Reply:
This doesn't raise a diagnostic, should it?

Reply:
Yeah it should be raising a diagnostic, it looks like the check being on line 228 was too early because `resp.TargetIdentity` is not set at that point (it's at `moveStateResp.TargetIdentity` when it comes back from your provider code that sets it).

Other RPCs just immediately set `resp.Identity` after the RPC, but if you want to move that piece of logic down lower we can also do that!

I'll make a comment suggestion 👍🏻